8253600: G1: Fully support pinned regions for full gc

tschatzl opened the pull request:
Hi all,

  can I get reviews for this change that implements "proper" support for pinned regions in the G1 full collector?

By proper I mean that at the end of gc, pinned regions contain the correct TAMS and bitmap markings under the TAMS so that dead objects within them are supported?

Currently all (pinned) regions have their TAMS set to bottom() and their bitmap above TAMS cleared (at least logically :) ). This works as long objects within these regions can't be dead as it is the case now:
- humongous regions are either live or fully reclaimed.
- all other pinned regions are archive regions at the moment that are always treated as fully live (and do not contain dead objects).

This change is a requirement for fixing JDK-8253081 as some earlier change made it possible to have dead objects within open archive regions. It also enables supporting removal of gclocker use for g1, i.e. using region pinning.

Based on the PR#808 (https://github.com/openjdk/jdk/pull/808).

Testing: tier1-8, testing with prototype for region pinning, testing with prototype for JDK-8253081.
Performance testing: no regressions

Some comments for questions that might come up during review:

- how does this work with the bitmaps now:
  - at start of full gc the next bitmap is cleared
  - full gc marks the next bitmap
  - for all pinned regions, keep TAMS and top() (*), otherwise set TAMS to bottom
  - swap bitmaps
  - clear next bitmap for next marking

(*) this means that from a usage POV pinned regions are considered full. This is inaccurate, but sufficient: full gc clears all remembered sets anyway, so we do not need that information for gc efficiency purposes anyway to evacuate later. The next marking before old gen evacuation will update it to the correct values anyway. G1 does not support allocation into "holes" in pinned regions that can be open archive only at this time too, so there is no need to be more exact.

- use of a region attribute table for phase 2+ only: compared to before we need fast access to information whether a given reference goes into a pinned region (as opposed to an archive region) wrt to adjusting that pointer to avoid doing work for these references.

Phase 1 marking could have used this information for the do-we-need-to-preserve-the-mark check too: however this would have required g1 to add an extra another pass over all regions to update that. This seemed slower than just checking this information "more slowly" for the objects that need mark preservation. Tests showed that this is the case for <0.00% (yeah, these references that need mark preservation are rounding errors in cases it matters) of overall references, so I did not add that pass.
(Additionally g1 full gc is a last-ditch effort, and while marking takes a significant time, it does not completely dominate it).

I.e. the second clause in the condition of this hunk is intentionally slower than could be:
```
@@ -52,7 +52,9 @@ inline bool G1FullGCMarker::mark_object(oop obj) {
  // Marked by us, preserve if needed.
  markWord mark = obj->mark();
  if (obj->mark_must_be_preserved(mark) &&
      // It is not necessary to preserve marks for objects in pinned regions because
      // we do not change their headers (i.e. forward them).
      !G1CollectedHeap::heap()->heap_region_containing(obj)->is_pinned()) {
    preserved_stack()->push(obj, mark);
  }
```
- there is no code yet that checks for empty pinned regions yet. Only JDK-8253081 introduces that because still all contents of all archive regions are live forever.

Also please note that the 51b297b change is from the #808 change.

Thanks,
  Thomas

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253600](https://bugs.openjdk.java.net/browse/JDK-8253600): G1: Fully support pinned regions for full gc


### Reviewers
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to 15db3776c22633c8bfb41f63ce0db5ebe3a9697e
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Author)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/824/head:pull/824`
`$ git checkout pull/824`

kstefanj commented on src/hotspot/share/gc/g1/heapRegion.hpp:
```diff
@@ -165,16 +165,12 @@ class HeapRegion : public CHeapObj<mtGC> {
-  void complete_compaction();
+  // Update heap region to be consistent after Full GC compaction for non-pinned
+  // regions.
+  void non_pinned_complete_compaction();
+  void pinned_complete_compaction();
+  void complete_compaction_common();
```
What do you think about using these names instead:
```suggestion
  void complete_compaction();
  void reset_pinned();
  void reset_after_full();
```

An other alternative would also to be to use `reset_after_compaction()` instead of `complete_compaction()`.

kstefanj commented on src/hotspot/share/gc/g1/g1FullCollector.hpp:
```diff
@@ -52,6 +53,44 @@ class G1FullGCSubjectToDiscoveryClosure: public BoolObjectClosure {
   }
 };
 
+// This table is used to store some per-region attributes needed during collection.
+class G1FullGCHeapRegionAttrBiasedMappedArray : public G1BiasedMappedArray<uint8_t> {
```
I would like this class to be moved to its own file.

kstefanj reviewed (changes requested):
Nice change Thomas, mostly just small comments.

tschatzl commented:
The new change (apart from the merge) should fix all the concerns @kstefanj mentioned.

In addition to that I looked into completely moving the G1ArchiveAllocator::_archive_region_map into the G1FullCollector::_region_attr_table as they serve the same purpose.

Other than in full gc, G1ArchiveAllocator::_archive_region_map is only used in non-perf critical code (object dumping, asserts), so the replacements should be more than fast enough.

The only drawback is that this adds a new method in CollectedHeap as a start for CDS archive support for other collectors which is thought of. More is needed, but that's the minimum to replace functionality previously provided by G1ArchiveAllocator::_archive_region_map at this time.

Its default implementation simply returns false if asked whether a given object on the heap is an archive object.

   virtual bool is_archived_object(oop object) const { return false; }

kstefanj commented on src/hotspot/share/gc/g1/g1FullCollector.hpp:
```diff
@@ -90,6 +94,11 @@ class G1FullCollector : StackObj {
 
+  void update_attribute_table(HeapRegion* hr);
+
+  bool is_in_pinned_or_closed(oop obj) const { return _region_attr_table.is_pinned_or_closed(cast_from_oop<HeapWord*>(obj)); }
+  bool is_in_closed(oop obj) const { return _region_attr_table.is_closed_archive(cast_from_oop<HeapWord*>(obj)); }
+
```
Can't decide if it's worth adding a inline.hpp for those. Just throwing it out there in case you thought about it as well :)

kstefanj reviewed (commented):
Thanks Thomas for addressing my concerns around the archive-map and the new region-attr overlap. I think this looks much better. 

Just a few additional comments.

tschatzl commented:
- I'll file a CR for that investigation. In my recent measurements of a different full-region closure we are talking about something < 0.0x ms.
- added an `is_in_pinned` method, and since now we've got three not-completely-trivial methods added a `.inline.hpp` file. I did not see need for moving the completely trivial getters returning just a member variable there, but tell me if I should move them too.
- moved "implementation" of `CollectedHeap::is_archived_object` to .cpp file

kstefanj reviewed (approved):
This looks great, thanks.

albertnetymk commented on src/hotspot/share/gc/g1/g1FullCollector.hpp:
```diff
@@ -71,6 +73,8 @@ class G1FullCollector : StackObj {
   G1FullGCSubjectToDiscoveryClosure _always_subject_to_discovery;
   ReferenceProcessorSubjectToDiscoveryMutator _is_subject_mutator;
 
+  G1FullGCHeapRegionAttr _region_attr_table;
```
I don't really see the point of this auxiliary data structure. Why can't we just query the underlying region for its type, pinned, open/close archive?

tschatzl replied:
The only reason is performance:

Consider memory access with that auxiliary data structure: it is a single dereference/memory load on a very dense data structure.

Directly querying the HeapRegion* is two dependent dereferences (first getting the HeapRegion*, then to the HeapRegion, then accessing the member) on a fairly large data structure (HeapRegion).

Since these loads are done for every reference (millions of times) the second is much slower, at best filling the cache with useless data as (roughly) a HeapRegion is around a cache line in size (iirc).

Further, with the current encoding of the values in region attribute table, the closed-or-pinned check can be done with a single check instead of some disjunction (ie. region->is_pinned() || region->is_closed(), although it is true that closed regions are always pinned), so saving even more code (and branch locations).

albertnetymk replied:
I see; thank you for the explanation. Please add some comments around this code.

albertnetymk reviewed (changes requested):
Thank you for the code walk-through. Some further questions/comments.

tschatzl commented on src/hotspot/share/gc/g1/g1FullGCHeapRegionAttr.hpp:
```diff
@@ -0,0 +1,74 @@
+
+  bool is_pinned_or_closed(HeapWord* obj) const {
+    assert(!is_invalid(obj), "not initialized yet");
+    return get_by_address(obj) >= Pinned;
```
Okay, will do.

tschatzl commented on src/hotspot/share/gc/g1/g1FullGCHeapRegionAttr.hpp:
```diff
@@ -0,0 +1,74 @@
+// This table is used to store some per-region attributes needed during full
+// collection.
+class G1FullGCHeapRegionAttr : public G1BiasedMappedArray<uint8_t> {
+  static const uint8_t Normal = 0;
+  static const uint8_t Pinned = 1;
+  static const uint8_t ClosedArchive = 2;
+
+  static const uint8_t Invalid = 255;
```
I do not understand what "Prior to this PR the default value is 0" means, can you clarify? The default value for the biased arrays must always be defined by the child classes overriding default_value() so there does not seem to be a pre-defined "prior" value.

Other than that the reason for using 255 is easier discrimination from the valid values in the debugger, just like hotspot uses something like 0xdeadbeef for various kinds of invalid memory.

There is no difference for the compiler for having 0 or 255 except for maybe the part about initializing the default value (on probably older x86 processors).

However compared to the overhead of setting up an efficient byte-wise `clear()` method is much larger than in this case a difference between `xor reg, reg` vs. `movsx reg, -1` (e.g. for x86, the former is/having been a common trick to quickly clear registers). That pales in comparison to the actual work.
Some architectures (ppc, sparc, iirc likely some more) have some special "clear cache line" instruction which may be used to clear memory "quickly" to zero, but it is not used since they often come with large caveats.

Otherwise the clearing of the this small table (one byte per region) is a one-time cost at startup of a full gc operations that may take billions of cpu cycles.

It's not worth thinking about overhead of the clearing in this case imho.

albertnetymk replied:
> can you clarify?

In the original code, `uint8_t default_value() const { return NoArchive; }`, where `NoArchive = 0`.

>  It's not worth thinking about overhead of the clearing in this case imho.

OK, maybe it's not perf sensitive.

albertnetymk reviewed (approved):
`G1FullGCHeapRegionAttr::is_pinned` and `HeapRegion::is_pinned` share the same name, but have different meanings. I wonder if it's possible to rename the former into sth more precise. Maybe `G1FullGCHeapRegionAttr::should_not_be_relocated`, or the version without negation `G1FullGCHeapRegionAttr::should_relocate`?

albertnetymk reviewed (approved):
Thanks to Thomas' offline explanation, my previous comment is incorrect, due to my erroneous understanding of the enum in `HeapRegionType`.

Thank you for new updates. This PR is good to go.

tschatzl commented:
Thanks @kstefanj  @albertnetymk for your reviews.

/integrate
